### DISCUSSION TITLE
How can I effectively use `Show` TypeClass?

### DISCUSSION THREAD

**Discussion by SHND** [gcanti/fp-ts#1934] [Q&A] [answered]
We have a `Show<A>` TypeClass in fp-ts. In the ideal situation when I want to get the string representation of Option, I want to simply call `show` method on the Show Option instance. Something like: `optionShow.show(some(12))`. But in real world our data may be `Option<List<number>>` or basically `Option<A>`. The only thing we need is that `A` has an instance of Show TypeClass. Then we can recursively and abstractly define Show instance for Option.

Can we define an instance of Show for Option like that in fp-ts? If not how should I think and organize code for Show instances?

**Comment by samhh** [comment, marked as answer]
I believe you're describing `getShow`: https://gcanti.github.io/fp-ts/modules/Option.ts.html#getshow

**Comment by SHND** [comment]
Amazing, thank you @samhh

In case anyone else found this thread, this is how recursively we can do this with `getShow`:

```ts
interface Show<A> {
  readonly show: (a: A) => string
}

const getShowNumber = (): Show<number> => ({
  show: (x) => `${x}`
})

const getShowOption = <A>(S: Show<A>): Show<Option<A>> => ({
  show: (ma) => (isNone(ma) ? 'none' : `some(${S.show(ma.value)})`)
})

const getShowList = <A>(S: Show<A>): Show<List<A>> => ({
  show: (ma) => {
    function _show(xs: List<A>): string {
      return isNil(xs) ? 'nil' : xs.head + ',' + _show(xs.tail);
    }

    return 'list(' + _show(ma) + ')'
  }
})

const data4 = some(cons(1, cons(2, cons(3, nil))));
const showOptionListNumber = getShowOption(getShowList(getShowNumber()))
const output4 = showOptionListNumber.show(data4)

console.log(output4)   // some(list(1,2,3,nil))
```
